### PR TITLE
Build speed improvements

### PR DESCRIPTION
**Pull request by Therzok**
This is the first part of fixing the build so it's snappy.

Right now, projects cause ResolveAssemblyReferences to be re-evaluated a lot of times and this just works towards fixing NOP builds to not compile anything.

Will add a CI based test that verifies that `make` does not contain targets we should be skipping.

TODO:

* [ ] Integrate with CI so we assert that things don't get rebuilt - maybe integrate a binlog parser and some msbuild task that automates this process
* [x] ~Fix md-addins~ - Will do in a later step to unblock this
* [x] Figure out why RAR takes a long time - Seems like nugets which contain references to shims happen to cause issues with resolving assemblies from the GAC. System.Runtime versions from nugets do not match the versions in the Mono GAC, so mono ends up probing a lot of directories trying to find the right versions.
* [ ] Remove tests local copies of MD assemblies, they should be loaded from mdtool/vstool
* [ ] Revert monodoc change for now.

Edit:
# What is this mastodon PR?

This PR tries to fix some common problems we have in how our projects reference eachother and nugets, avoiding issues with incremental builds.

It tries to consolidate nuget versions into props files and only MonoDevelop.Core has to do the NuGet restore step.

It also fixes issues with mismatched versions by adding some binding redirects.

Unifies assembly references and only local copies from where it should be.

# What do we gain?

Better incrementality.

We removed the system dependency on monodoc, we now get it from nuget, but it's a bit hacky:
https://github.com/mono/api-doc-tools/issues/229
https://github.com/mono/api-doc-tools/issues/228
https://github.com/mono/api-doc-tools/issues/227

# There are some missing changes

I reverted the changes in F# until we have a consensus about switching from Paket to this model.

I haven't opened PRs yet to mono-addins, debugger-libs and NRefactory

# Current issues

PackageReference would solve a lot of our issues, and we'd only need to have versions defined, and not manual references to different assemblies, right now, I used what nuget added to the csproj files.

### CODE REVIEW

File: main/external/Makefile.am [hunk 1]
```diff
@@ -53,4 +56,3 @@ dist-clean:
 		fi \
 	done
 
-.PHONY: Xamarin.Mac.dll
```
Comment (Therzok) [inline main/external/Makefile.am]: This was causing Xamarin.Mac to always be copied from `make`.

**Comment by Therzok** [comment]
Working towards this: https://github.com/mono/monodevelop/issues/4521

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj [hunk 2]
```diff
@@ -190,7 +190,6 @@
     <Reference Include="System.Runtime.Serialization" />
     <Reference Include="System.Threading.Tasks.Dataflow, Version=4.6.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
       <HintPath>..\..\..\packages\System.Threading.Tasks.Dataflow.4.7.0\lib\netstandard1.0\System.Threading.Tasks.Dataflow.dll</HintPath>
-      <Private>True</Private>
```
Comment (KirillOsenkov) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj]: >True [](start = 15, length = 4)

Maybe an explicit False just in case?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj]: Sure! But it's not appearing as double right in binlog anymore

**Comment by Therzok** [comment]
So, we can't get a SharpZipLib that works for monodoc and AndroidSDK manager. Maybe we can get upstream monodoc to work with the stripped version.

**Comment by KirillOsenkov** [comment]
Looking great so far! :+1:

**Comment by Therzok** [comment]
I'm considering reverting the monodoc change for now and file new issues for migrating monodoc and mono.posix gac usage to nuget.

**Comment by Therzok** [comment]
Temporary monodoc solution: create a locally nupkg that we publish via xampubbot or myget.

Will redo binding redirect fixes based on what Windows reports and I'll file bugs to mono/msbuild about the inconsistencies on mac.

**Comment by Therzok** [comment]
I'll split this PR into atomic PRs. Better than reviewing one massive PR

**Comment by Therzok** [comment]
Closing. We already had a migration to PackageReference. If we want to fix rebuilding again, we need to redo the PR completely.

**Pull request closed without merging**